File: Estruturais/Flyweight.java
package Estruturais;

import java.util.HashMap;
import java.util.Map;

// Objeto Flyweight compartilhado
class TreeType {
    private String name;
    private String color;
    private String texture;

    public TreeType(String name, String color, String texture) {
        this.name = name;
        this.color = color;
        this.texture = texture;
    }

    public void draw(int x, int y) {
        System.out.println("Desenhando a árvore '" + name + "' na posição (" + x + ", " + y + ") com cor '" + color + "' e textura '" + texture + "'.");
    }
}

// Fábrica de Flyweight
class TreeFactory {
    private static final Map<String, TreeType> treeTypes = new HashMap<>();

    public static TreeType getTreeType(String name, String color, String texture) {
        String key = name + color + texture;
        if (!treeTypes.containsKey(key)) {
            treeTypes.put(key, new TreeType(name, color, texture));
        }
        return treeTypes.get(key);
    }
}

// Uso do Flyweight
class Tree {
    private int x, y;
    private TreeType treeType;

    public Tree(int x, int y, TreeType treeType) {
        this.x = x;
        this.y = y;
        this.treeType = treeType;
    }

    public void draw() {
        treeType.draw(x, y);
    }
}

// Exemplo de uso
public class Flyweight {
    public static void main(String[] args) {
        Tree tree1 = new Tree(10, 20, TreeFactory.getTreeType("Pine", "Green", "Smooth"));
        Tree tree2 = new Tree(30, 40, TreeFactory.getTreeType("Pine", "Green", "Smooth"));
        Tree tree3 = new Tree(15, 25, TreeFactory.getTreeType("Oak", "Brown", "Rough"));

        tree1.draw();
        tree2.draw();
        tree3.draw();
    }
}

File: Estruturais/Flyweight.py
# Objeto Flyweight compartilhado
class TreeType:
    def __init__(self, name, color, texture):
        self.name = name
        self.color = color
        self.texture = texture

    def draw(self, x, y):
        print(f"Desenhando a árvore '{self.name}' na posição ({x}, {y}) com cor '{self.color}' e textura '{self.texture}'.")

# Fábrica de Flyweight
class TreeFactory:
    _tree_types = {}

    @classmethod
    def get_tree_type(cls, name, color, texture):
        if (name, color, texture) not in cls._tree_types:
            cls._tree_types[(name, color, texture)] = TreeType(name, color, texture)
        return cls._tree_types[(name, color, texture)]

# Uso do Flyweight
class Tree:
    def __init__(self, x, y, tree_type):
        self.x = x
        self.y = y
        self.tree_type = tree_type

    def draw(self):
        self.tree_type.draw(self.x, self.y)

# Exemplo de uso
tree1 = Tree(10, 20, TreeFactory.get_tree_type("Pine", "Green", "Smooth"))
tree2 = Tree(30, 40, TreeFactory.get_tree_type("Pine", "Green", "Smooth"))
tree3 = Tree(15, 25, TreeFactory.get_tree_type("Oak", "Brown", "Rough"))

tree1.draw()
tree2.draw()
tree3.draw()
